### PR TITLE
Use QueryPages

### PR DESCRIPTION
**Pull request by tlinton**
Hi,

I hope you accept PRs that come out of the blue like this.

I have been using the Timestream adapter and noticed that it does not support paging when reading large queries, so I added that.

This PR also adds a diagnostic message if a measure contains illegal characters, which is something I had problems with and it took me a while to understand what was going on.

Thanks,
Tobias Linton

### CODE REVIEW

**Comment by dpattmann** [comment]
Hi @tlinton,

thank you for the PR!

Can you please explain the issue with Query and why you replaced it with QueryPages?

File: timestream.go [hunk 1]
```diff
@@ -211,15 +227,26 @@ func (t TimeStreamAdapter) runReadRequestQuery(q *prompb.Query) (result prompb.Q
...
-		QueryString: &task.query,
-	})
+	var timeSeries []*prompb.TimeSeries
+	var innerErr error
+	err = t.QueryPages(
+		&timestreamquery.QueryInput{
+			QueryString: &task.query,
+		},
+		func(output *timestreamquery.QueryOutput, lastPage bool) (continueIterating bool) {
+			continueIterating = true
```
Comment (dpattmann) [inline timestream.go:237]: I'm not sure if this function should always true.

**Comment by tlinton** [comment]
Hi @dpattmann,

> Hi @tlinton,
> 
> thank you for the PR!
> 
> Can you please explain the issue with Query and why you replaced it with QueryPages?

If the query result is large, TimeStream will paginate it. The client is supposed to request the different pages by issuing new requests, and setting `QueryInput.NextToken` to the value of `QueryOutput.NextToken` from the previous response. `QueryPages` manages the token handling automatically, if the response is paginated it will send the multiple requests and send the result to the function that was provided as argument.

When a response is paginated, the first response contains 0 rows. This is how I noticed this, some of my requests did not return any data.

File: timestream.go (hunk 1 above)
Comment (tlinton) [reply timestream.go:237]: This is not very well documented, but my understanding is that it only needs to return false if it wants to exit before all pages of the response have been read. As far as I can tell this works, but setting `continueIterating = !lastPage` would make it nicer. I'll do that and see what happens.

Comment (tlinton) [reply timestream.go]: Setting `continueIterating = !lastPage` works fine!

File: timestream.go [hunk 2]
```diff
@@ -211,15 +227,26 @@ func (t TimeStreamAdapter) runReadRequestQuery(q *prompb.Query) (result prompb.Q
...
+	var innerErr error
+	err = t.QueryPages(
+		&timestreamquery.QueryInput{
+			QueryString: &task.query,
+		},
+		func(output *timestreamquery.QueryOutput, lastPage bool) (continueIterating bool) {
+			continueIterating = !lastPage
+			timeSeries, innerErr = t.handleQueryResult(output, timeSeries, task.measureName)
+			return
+		},
```
Comment (dpattmann) [inline timestream.go:240]: Please return ```!lastPage``` directly and remove the ```continueIterating``` variable.

File: timestream.go [hunk 3]
```diff
@@ -140,6 +140,19 @@ func (t TimeStreamAdapter) Write(req *prompb.WriteRequest) (err error) {
 	return
 }
 
+func allCharactersValid(str string) bool {
```
Comment (dpattmann) [inline timestream.go:143]: Please make this a TimeStreamAdapter methode

**Comment by dpattmann** [comment]
Thanks!

**Comment by tlinton** [comment]
> Thanks!

Thank you!

**Pull request merged**